### PR TITLE
pubsub: Add subscriptions

### PR DESCRIPTION
**Pull request by rakyll**
/cc @broady

### CODE REVIEW

File: pubsub/subscriptions/main.go [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+	}
+	fmt.Printf("Created subscription: %v\n", sub)
+	// [END create_subscription]
+	return nil
+}
+
+func delete(c *pubsub.Client, name string) error {
+	ctx := context.Background()
+	// [START delete_subscription]
+	sub := c.Subscription(name)
```
Comment (broady) [inline pubsub/subscriptions/main.go]: s/c/client/g

File: pubsub/subscriptions/main.go [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+	if err := sub.Delete(ctx); err != nil {
+		return fmt.Errorf("Failed to delete the subscription: %v\n", err)
```
Comment (broady) [inline pubsub/subscriptions/main.go]: I don't think we need to wrap the errors for the samples. Also, don't put newlines at the end of wrapper errors.

File: pubsub/subscriptions/main.go [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+	it, err := sub.Pull(ctx)
+	if err != nil {
+		return fmt.Errorf("Failed to pull from subsription: %v", err)
+	}
+	defer it.Stop()
+
+	// Consume 10 messages.
+	for i := 0; i < n; i++ {
+		msg, err := it.Next()
+		if err == pubsub.Done {
```
Comment (broady) [inline pubsub/subscriptions/main.go]: pubsub.Done will never be returned here.

File: pubsub/subscriptions/main.go [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+	// [START pull_messages]
+	sub := c.Subscription(name)
+	it, err := sub.Pull(ctx)
+	if err != nil {
+		return fmt.Errorf("Failed to pull from subsription: %v", err)
+	}
+	defer it.Stop()
+
+	// Consume 10 messages.
+	for i := 0; i < n; i++ {
```
Comment (broady) [inline pubsub/subscriptions/main.go]: n is outside the scope of the region tag

File: pubsub/subscriptions/main.go (hunk 1 above)
Comment (rakyll) [inline pubsub/subscriptions/main.go]: Done

File: pubsub/subscriptions/main.go (hunk 2 above)
Comment (rakyll) [inline pubsub/subscriptions/main.go]: Done

File: pubsub/subscriptions/main.go (hunk 3 above)
Comment (rakyll) [inline pubsub/subscriptions/main.go]: Unless another client is also consuming the topic.

We should keep it for having a reference how to use the iterator.

File: pubsub/subscriptions/main.go (hunk 4 above)
Comment (rakyll) [inline pubsub/subscriptions/main.go]: Hardcoded. Not ideal but these are snippets.

**Comment by rakyll** [comment]
PTAL

Also added tests.

File: pubsub/subscriptions/main.go (hunk 3 above)
Comment (broady) [inline pubsub/subscriptions/main.go]: It's only returned after Stop is called. It's actually pretty strange that `pubsub.Done` exists at all.

File: pubsub/subscriptions/main.go [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+		fmt.Printf("%v\n", sub.Name())
+	}
+
+	const topic = "example-topic"
+	// Create a topic to subscribe to.
+	t, err := client.NewTopic(ctx, topic)
+	if err != nil {
+		log.Fatalf("Failed to create the topic: %v", err)
+	}
+	defer t.Delete(ctx) // cleanup when finished using t.
```
Comment (broady) [inline pubsub/subscriptions/main.go]: defer is not called when `log.Fatal` is called

**Comment by broady** [comment]
LGTM

**Pull request merged**